src/main/java/com/google/opensesame/projects/ProjectQuery.java

```diff
@@ -0,0 +1,118 @@
+package com.google.opensesame.projects;
+
+import static com.googlecode.objectify.ObjectifyService.ofy;
+
+import com.google.opensesame.util.ServletValidationException;
+import com.googlecode.objectify.annotation.Index;
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Helper functions for querying the projects Datastore. */
+public class ProjectQuery {
+  public static final String FILTER_QUERY_PARAM = "filter";
+
+  /**
+   * The fields of ProjectEntity that have the @Index Objectify annotation and can be used for
+   * sorting and filtering in Datastore.
+   */
+  public static List<Field> queryableFields;
+
+  static {
+    Field[] fields = ProjectEntity.class.getDeclaredFields();
+    queryableFields = new ArrayList<Field>();
+    for (Field field : fields) {
+      if (field.isAnnotationPresent(Index.class)) {
+        queryableFields.add(field);
+      }
+    }
+  }
+
+  /**
+   * Gets the query filters from a servlet request. Returns null if any errors are encountered while
+   * parsing the query filters. See {@link com.google.opensesame.projects.QueryFilter.fromString}
+   * for more information on how QueryFilters are parsed.
+   *
+   * @param HttpServletRequest The request to get the filter queries from.
+   * @param HttpServletResponse The servlet response to send errors to.
+   * @return Returns the list of QueryFilters or null if an error was encountered while parsing.
+   * @throws ServletValidationException
+   */
+  private static List<QueryFilter> getQueryFiltersFromRequest(HttpServletRequest request)
+      throws IOException, ServletValidationException {
+    String[] filterRequests = request.getParameterValues(FILTER_QUERY_PARAM);
+    if (filterRequests == null) {
+      return Arrays.asList();
+    }
+
+    List<QueryFilter> queryFilters = new ArrayList<QueryFilter>();
+    for (String filterRequest : filterRequests) {
+      QueryFilter queryFilter = QueryFilter.fromString(filterRequest);
+      queryFilters.add(queryFilter);
+    }
+
+    return queryFilters;
+  }
+
+  /**
+   * Queries for ProjectEntities in the datastore based on a servlet request.
+   *
+   * @param request The servlet request defining the query.
+   * @param response The servlet response to send errors to.
+   * @return Returns a collection of ProjectEntities from the query or null if the query was
```

Comment:
does this function return null or does it just raise an error?

Reply:
Apologies, this is due to me forgetting to update the JavaDoc after moving to using exceptions. This will be updated!